File: day11/11.c
#include"../a.h"
C g[140][140],h[140],v[140];T q[500][2];L h2[140],v2[140],gn,d;
L cab(L i,L j)_(abs(h2[q[i][0]]-h2[q[j][0]])+abs(v2[q[i][1]]-v2[q[j][1]]))
V scal(L s){
  L x=0,y=0;Fi(140,h2[i]=x+=!h[i]?s:1,v2[i]=y+=!v[i]?s:1)
  d=0;Fi(gn,Fj(i,d+=cab(i,j)))printf("%ld\n",d);}
I main() {
    FILE*f=fopen("input.txt","r");
    Fi(140,Fj(140,g[i][j]=fgetc(f)=='#')fgetc(f));
    Fi(140,Fj(140,h[i]|=g[i][j],v[i]|=g[j][i]))
    Fi(140,Fj(140,IF(g[i][j],q[gn][0]=i;q[gn][1]=j;gn++)))
    scal(2);scal(1E6);}

File: day11/11.hs
import Data.Array (Array)
import Data.List (transpose)
import Data.Bifunctor ( Bifunctor(bimap) )

readData :: FilePath -> IO [[Bool]]
readData file = do
  contents <- readFile file
  return $ map (map (== '#')) (lines contents)

empty :: [[Bool]] -> ([Bool], [Bool])
empty grid = (map or grid, map or (transpose grid))

locations :: [[Bool]] -> [(Int, Int)]
locations grid = concat $ zipWith (\x y -> map (\z -> (x, z)) y) [0..] (map (map fst . filter snd . zip [0..]) grid)

cab :: (Int, Int) -> (Int, Int) -> Int
cab (x1, y1) (x2, y2) = abs (x1 - x2) + abs (y1 - y2)

scal :: [[Bool]] -> Int -> Int
scal grid factor = sum [cab (locs !! i) (locs !! j) | i <- [0..(length locs - 1)], j <- [0..i]]
    where (emptyRows, emptyCols) = empty grid
          rows = scan emptyRows
          cols = scan emptyCols
          locs = map (bimap (rows !!) (cols !!)) (locations grid)
          scan x = map pred (tail $ scanl (\acc x -> acc + if x then 1 else factor) 0 x)

main :: IO ()
main = do
  grid <- readData "input.txt"
  putStrLn $ "Part 1: " ++ show (scal grid 2)
  putStrLn $ "Part 2: " ++ show (scal grid 1000000)